### DISCUSSION TITLE
Feature request: Expose HTTP request metrics

### DISCUSSION THREAD

**Discussion by juliusv** [basecamp/kamal-proxy#101] [Ideas and Issue Triage] [answered]
As far as I can see, kamal-proxy does not yet expose metrics about the requests it serves. It would be really nice if it tracked e.g. Prometheus metrics for things like:

* Request counts (including status code label dimension)
* Request duration histogram
* Request size histogram
* Response size histogram
* Current number of in-flight requests
* Etc.

**Comment by northeastprince** [comment]
Request metrics seen more like a responsibility for the apps themselves.

**Comment by juliusv** [comment]
That's ok, though it's pretty common to have them in both places. Usually apps are instrumented, and proxies / LBs like nginx, HAProxy, etc. also track metrics about requests going to their services and backends.

**Comment by gouthamve** [comment]
Can I request that this be reopened? Lets say the application is crashing and kamal-proxy is returning `5xx`, we cannot alert on this because the app itself is down and `kamal-proxy` has no metrics.

Further, there can be cases where the `proxy` latency is different from `app` latencies, and the `proxy` latencies are much closer to what users see. It's best to also expose metrics from the `proxy` to diagnose this issue.

**Comment by polarctos** [comment, marked as answer]
Metrics via HTTP are now available, added in: https://github.com/basecamp/kamal-proxy/pull/126 available in [v0.9.1](https://github.com/basecamp/kamal-proxy/releases/tag/v0.9.1)

https://github.com/basecamp/kamal-proxy/blob/950b2525d9593fe96a1353f8655426f0b2c65284/internal/cmd/run.go#L30

You can add e.g. `--metrics-port=9001` to enable it.

Whats not there yet is metrics via HTTPS port.

I was wondering about adding an additional flag `--metrics-https=true` to toggle setting the TLS config on the metrics server too.

**Reply by polarctos** [reply]
Opened a PR for HTTPS: https://github.com/basecamp/kamal-proxy/pull/161

**Reply by kevinmcconnell** [reply]
@polarctos thanks for answering this!

I had some questions about the HTTPS option that I'd like to understand better. I've moved [that discussion](https://github.com/basecamp/kamal-proxy/pull/161#pullrequestreview-3237984879) to the PR.
